### PR TITLE
Candidate for #689

### PR DESCRIPTION
**Pull request by csgordon**
This is a candidate fix for #689.  It appears the assertion was too strong, because apparently none of the work on the ECOOP paper encountered anonymous inner classes with non-trivial (i.e., method-invoking) field initializers, which don't get visited in the context of an enclosing method structure.

A test for this is necessary as well before I close the issue, but I want this to go through CI to maybe unstick things for Monday.

### CODE REVIEW

File: checker/src/org/checkerframework/checker/guieffect/GuiEffectVisitor.java [hunk 1]
```diff
@@ -156,7 +156,9 @@ public Void visitMethodInvocation(MethodInvocationTree node, Void p) {
         }
 
         Effect callerEffect = atypeFactory.getDeclaredEffect(callerElt);
-        assert (callerEffect.equals(effStack.peek()));
+        // Field initializers inside anonymous inner classes show up with a null current-method ---
+        // the traversal goes straight from the class to the initializer.
+        assert (currentMethods.peek() == null || callerEffect.equals(effStack.peek()));
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/guieffect/GuiEffectVisitor.java]: Would `currentMethods.empty()` be simpler for the first condition? I hope we don't expect a null value on the stack.

**Comment by wmdietl** [comment]
Is there a minimal test case that you can also add?
I'll look at your comment in the issue tomorrow.
Thansk!

File: checker/src/org/checkerframework/checker/guieffect/GuiEffectVisitor.java (hunk 1 above)
Comment (csgordon) [inline checker/src/org/checkerframework/checker/guieffect/GuiEffectVisitor.java]: We do; ```currentMethods``` is a stack of ```MethodTree```s, where ```null``` is used as a sentinel to indicate that there is no surrounding method context.  Previously I thought this only applied to static initializers (which is why the sentinel exists in the first place), but apparently instance initializers also show up in this context.

If there's a more idiomatic way to handle the sentinel (some kind of non-null dummy method tree) I'm open to a bit of refactoring.

**Comment by csgordon** [comment]
There is, I just haven't had a chance to sit down and work through the plumbing yet.

**Comment by csgordon** [comment]
@wmdietl there is now a test

**Pull request merged**